File Upload Exercise Assessment

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This is the last third PR related to adding new exercise type, File Upload Exercise, into Artemis. This PR focuses on allowing tutor to assess student submissions and for students to see the results of assessments.

### Description
<!-- Describe your changes in detail -->
Tutors are able to add feedback to submisisons. Tutors can add any number of feedback items, each with description and score. In this case `referencedFeedback` doesn't have any real reference. Also tutors can give general feedback for the submission. Additionally, tutor should not be able to get any data about the student
Students can check the feedback. Each feedback item will be shown separetely to the user. Additionally, user should not be able get any data about the assessor.
Students can also complain/request more feedback in the same way as in other exercises, tutors should be able to see the request and evaluate it.

**NOTE**: I moved some of the code into generic classess, but there is still code duplication in server and client side code. Reducing code duplication will require to change modeling and text exercises and test their functionality. This is out of scope of this PR and will be tackled in a separate PR.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. For File Upload Exercise create a new submission
4. Go to Tutor Exercise Dashboard
5. Start assessment of the File Upload Exercise
6. Submit assessment
7. Check assessment as a student
8. Complain about assessment
9. Evaluate complaint as a tutor
10. Check complaint response as a student

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
#### File Upload Exercise in Tutor Exercise Dashboard 
![Screenshot_2019-09-23 Tutor Exercise Dashboard](https://user-images.githubusercontent.com/17262376/65433925-391b8200-de1e-11e9-8708-31acad4ddb58.png)
---
#### Assessment
![Screenshot_2019-09-24 File Upload Exercises](https://user-images.githubusercontent.com/17262376/65552219-2b502480-df24-11e9-8e20-dda1c111b8e9.png)
---
#### Student checks assessment
![Screenshot_2019-09-24 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/65553450-2fca0c80-df27-11e9-8e64-4ddecd88afaa.png)
---
#### Tutor evaluated complaint
![Screenshot_2019-09-25 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/65557038-23977c80-df32-11e9-9a0c-9232968c434f.png)
---
#### Student complaint was accepted
![Screenshot_2019-09-25 File Upload Exercises](https://user-images.githubusercontent.com/17262376/65555963-4e7fd180-df2e-11e9-91ce-b94153697803.png)

thilo-behnke commented on src/main/webapp/app/assessment-shared/assessment-detail/assessment-detail.component.ts:
```diff
@@ -26,7 +26,7 @@ export class TextAssessmentDetailComponent {
 
     public delete() {
-        const referencedText = this.block ? this.block.text : this.assessment.reference;
+        const referencedText = this.block ? this.block.text : this.assessment.reference ? this.assessment.reference : '';
```
Nit: I try to avoid nested if else, I would propose using boolean logic here:
`const referencedText = this.block && this.block.text || this.assessment.reference || '';`

In my opinion this is clearer when it comes to the different alternatives.

maxr96 replied:
I changed it to check if `referencedText` is null when I build `confirmationMessage`. So now it is not needed here

thilo-behnke commented on src/main/webapp/app/entities/file-upload-assessment/file-upload-assessment.service.ts:
```diff
@@ -0,0 +1,43 @@
+
+    saveAssessment(feedbacks: Feedback[], submissionId: number, submit = false): Observable<Result> {
+        let url = `${this.resourceUrl}/file-upload-submissions/${submissionId}/feedback`;
+        if (submit) {
```
Nit: It would be nice to write a small util for this that appends params to the URL.

maxr96 replied:
Added `buildUrlWithParams` to `global.utils` file

thilo-behnke commented on src/main/webapp/app/file-upload-assessment/file-upload-assessment.component.html:
```diff
@@ -0,0 +1,75 @@
+            <jhi-resizable-instructions
+                class="resizable-submission card"
+                id="instructions"
+                [ngStyle]="{ 'width.px': 400, 'min-width.px': 400 }"
```
I would prefer using a component scss file here, the styling can be hard to find when issues come up :see_no_evil:

maxr96 replied:
You are right!

thilo-behnke commented on src/main/webapp/app/file-upload-assessment/file-upload-assessment.component.ts:
```diff
@@ -0,0 +1,482 @@
+            );
+    }
+
+    private handleReceivedSubmission(submission: FileUploadSubmission): void {
```
Very generic method name. What does it mean that the receivedSubmission is handled?
I also wonder why we call `loadFeedbacks` both in `loadSubmissions` and here, even though `loadSubmissions` already calls it :thinking: This has an indication of spaghetti code to me.

maxr96 replied:
You are right about `loadFeedbacks`, it should be called only in `handleReceivedSubmission`. I changed the name `handleReceivedSubmission` to `initializePropertiesFromSubmission` to make it more clear that we initialize component properties from submission here

thilo-behnke commented on src/main/webapp/app/file-upload-assessment/file-upload-assessment.component.ts:
```diff
@@ -0,0 +1,482 @@
+            this.jhiAlertService.info('artemisApp.assessment.messages.lock');
+        }
+
+        this.formattedGradingInstructions = this.artemisMarkdown.htmlForMarkdown(this.exercise.gradingInstructions);
```
You could get rid of the attributes if you would use the `renderMarkdown` pipe in the template.

maxr96 replied:
True, looks cleaner now

thilo-behnke commented on src/main/webapp/app/file-upload-assessment/file-upload-assessment.component.ts:
```diff
@@ -0,0 +1,482 @@
+     * For the new submission to appear on the same page, the url has to be reloaded.
+     */
+    assessNextOptimal() {
+        if (this.exercise.type === ExerciseType.FILE_UPLOAD) {
```
Why do we need to check this here? Isn't the component exclusively used for file upload exercises?

maxr96 replied:
Yes, I removed the check

thilo-behnke commented on src/main/webapp/app/file-upload-assessment/file-upload-assessment.component.ts:
```diff
@@ -0,0 +1,482 @@
+    }
+
+    onSaveAssessment() {
+        this.fileUploadAssessmentsService.saveAssessment(this.assessments, this.submission!.id).subscribe(
```
Something I wonder in general here: There doesn't seem to be any user feedback while a REST call is pending. E.g. what if saveAssessment would take 10 seconds? Would the user understand the process?

maxr96 replied:
You are right that this was missing, I set `isLoading` flag here

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-result/file-upload-result.component.ts:
```diff
@@ -0,0 +1,24 @@
+        if (!result) {
+            return;
+        }
+        const groupedFeedback = partition(result.feedbacks, feedback => feedback.credits > 0);
```
Very nice!

You could improve it further:
`const [feedbackWithCredits, feedbackWithoutCredits] = partition(result.feedbacks, feedback => feedback.credits > 0);`

This naming helps to understand what you are partitioning.

maxr96 replied:
Good suggestion! Indeed this is more clear

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-result/file-upload-result.component.ts:
```diff
@@ -0,0 +1,24 @@
+        }
+        const groupedFeedback = partition(result.feedbacks, feedback => feedback.credits > 0);
+        this.feedbacks = groupedFeedback[0];
+        this.generalFeedback = groupedFeedback[1][0];
```
Is it a given that `groupedFeedback[1][0]` exists?

maxr96 replied:
No, it can be null, I explicitly added this to the type definition now

thilo-behnke replied:
:+1:

thilo-behnke commented on src/main/webapp/app/file-upload-submission/file-upload-submission.component.ts:
```diff
@@ -85,14 +93,30 @@ export class FileUploadSubmissionComponent implements OnInit, ComponentCanDeacti
+                        this.fileUploadAssessmentService.getAssessment(this.submission.id).subscribe((assessmentResult: Result) => {
+                            this.result = assessmentResult;
+                        });
+                        this.complaintService
```
This nested structure makes it hard to understand the code (https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming)).

With RxJS you can easily chain multiple steps on the same level:

```
service.get(abc).pipe(
  map(sum),
  tap(informClient),
  switchMap(() => nowGetOtherThing(def))
).subscribe();

maxr96 replied:
I simplified if statement by setting `this.result` beforehand and not making a huge `if else`. This also fixes the bug with complaint not loaded properly. Not sure if any RxJS changes are needed here now.

thilo-behnke commented on src/main/webapp/app/tutor-exercise-dashboard/tutor-exercise-dashboard.component.ts:
```diff
@@ -342,7 +342,7 @@ export class TutorExerciseDashboardComponent implements OnInit {
                 break;
             case ExerciseType.FILE_UPLOAD:
-                route = `/file-upload-exercise/${this.exercise.id}/submissions/${submission}/assessment`;
+                route = `/file-upload-exercise/${this.exercise.id}/submission/${submission}/assessment`;
```
Wasn't `submissions` correct here?

maxr96 replied:
For client routes we use singular, not plural. This will be also changed for modeling and text exercises later

thilo-behnke replied:
Ah, you are right!

thilo-behnke commented on src/main/webapp/i18n/de/assessment.json:
```diff
@@ -13,6 +14,21 @@
+                "confirmCancel": "Bist du sicher, dass du die Bewertung abbrechen möchtest? Dein aktueller Bewertungsfortschritt wird gelöscht und die Abgabe steht anderen Tutoren wieder zur Verfügung.",
+                "loadSubmissionFailed": "Laden der angefragten Abgabe ist fehlgeschlagen.",
+                "submitSuccessful": "Deine Bewertung wurde erfolgreich abgesendet!",
+                "lock": "Dieses Modell kann jetzt exklusiv nur noch von dir bearbeitet werden. Bitte bewerte dieses Modell bevor du ein neues öffnest.",
```
Is `Modell` the right term here? Isn't the assessment neutral to the exercise type?

maxr96 replied:
You are right here. I wanted to unify translations, but didn't see that this is inappropriate for file-upload assessment. Changed this

thilo-behnke commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,17 @@
+            "error": {
+                "invalidAssessments": "Deine Bewertung ist nicht gültig!",
+                "invalidScoreMustBeNumber": "Das Score-Feld muss eine Zahl sein und darf nicht leer sein!",
+                "invalidNeedScore": "Jedes Feedback-Element muss mindestens eine Punktzahl  enthalten."
```
```suggestion
                "invalidNeedScore": "Jedes Feedback-Element muss mindestens eine Punktzahl enthalten."
```

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadAssessmentResource.java:
```diff
@@ -0,0 +1,184 @@
+            return notFound();
+        }
+
+        if (!authCheckService.isUserAllowedToGetResult(exercise, participation, result)) {
```
Yes, it is not reachable for users that are not the owner of participation, even tutors/instructors.

thilo-behnke replied:
:+1:

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadAssessmentResource.java:
```diff
@@ -0,0 +1,184 @@
+
+    /**
+     * Get the result of the file upload submission with the given id. Returns a 403 Forbidden response if the user is not allowed to retrieve the assessment. The user is not allowed
+     * to retrieve the assessment if he is not a student of the corresponding course, the submission is not his submission, the result is not finished or the assessment due date of
```
I saw we are also using `he/she` or `his/her` and used it here as well

thilo-behnke replied:
:+1:

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -246,50 +253,46 @@ public FileUploadSubmissionResource(FileUploadSubmissionRepository fileUploadSub
 
-        if (!authCheckService.isAtLeastInstructorForExercise(fileUploadExercise)) {
-            participation.setStudent(null);
+        if (fileUploadSubmission.getResult() != null && !authCheckService.isAtLeastTeachingAssistantForExercise(fileUploadExercise)) {
```
In this comment we don't want to send information about the assessor, if the user that requested it is not at least tutor for exercise

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadAssessmentResource.java:
```diff
@@ -0,0 +1,184 @@
+        Result result = fileUploadAssessmentService.updateAssessmentAfterComplaint(fileUploadSubmission.getResult(), fileUploadExercise, assessmentUpdate);
+
+        // remove circular dependencies if the results of the participation are there
+        if (result.getParticipation() != null && Hibernate.isInitialized(result.getParticipation().getResults()) && result.getParticipation().getResults() != null) {
```
You are right here it is not required. It works properly without it. However, I am not sure then why it is used for Modeling Exercise

maxr96 commented:
> 
> 
> Looks solid 👍
> 
> **Found Bugs:**
> 
>     1. Major: The instructor actions are not showing up (compare to e.g. modeling exercise) in the exercise detail page. I had to test this feature by manually navigating to the routes:
> 
> 
> ![image](https://user-images.githubusercontent.com/28230611/65759210-b6930b00-e11a-11e9-8aa2-f5e4780f3077.png)
> 
>     1. Major: After a complaint was rejected and the student goes to `view result`, a bad request is executed: http://localhost:9000/api/complaints/result/null. This seems to be a timing issue in the component as it works in 40% of the cases (maybe two async calls running at the same time?)
> 
>     2. Not sure if as intended: I can't submit a second complaint on the same exercise:
> 
> 
> ![image](https://user-images.githubusercontent.com/28230611/65759404-273a2780-e11b-11e9-91dd-d817739ebbf2.png)

Thank you @thilo-behnke for your feedback. I worked on the issues you pointed out, the ones that was clear how to solve I resolved, for other issues I left a comment, so you can check if you agree with it. Regarding the bugs:
1. This is a very good find. I didn't think about this use case, so it was not implemented. Now I added `file-upload-dashboard` that will be shown when navigating to `Assessments`. I also updated the translations, since they are the same for modeling, text and file-upload exercises. The `Statistics` is still only available for Modeling exercise, but in the future we can also consider adding it to the file upload and text exercises.
2. As you correctly pointed out that was a timing issue, as I didn't have the variable, when calling the endpoint to get complaints. This is fixed now
3. This is intended, You can only complain or request more feedback once per exercise.

maxr96 commented:
> 
> 
> @maxr96 I can confirm that the other bugs are fixed 👍
> 
>     1. Now for me when I have created a complaint as a student, I can't react to it anymore as the instructor (shouldn't there be a text panel with a button on the right?)
> 
> 
> ![image](https://user-images.githubusercontent.com/28230611/65814558-a7c45b00-e1e3-11e9-88d7-9be5f5654992.png)
> 
>     1. Another thing where I'm not sure if as intended: Is it normal that I can assess my own submission?

Thank you for rechecking @thilo-behnke
1. Yes, you can't react to the complaint, if you are logged in as assessor. Only another tutor can react to the complaint. For more feedback requests it is vice versa.
2. In my opinion this is normal, since in real world usage it should not be the case that user is both student and tutor for the same course. And this is also how it works now for modeling and text exercises.

lovaiible commented on src/main/webapp/i18n/de/fileUploadSubmission.json:
```diff
@@ -16,7 +16,7 @@
             },
             "filePath": "File Path",
-            "feedbackResult": "Du hast {{ credits }} Credits für {{ part }} Teil erhalten",
+            "feedbackResult": "Du hast {{ credits }} Credits",
```
```suggestion
            "feedbackResult": "Du hast {{ credits }} Punkt(e)",
```

lovaiible commented on src/main/webapp/i18n/en/fileUploadSubmission.json:
```diff
@@ -16,8 +16,8 @@
             "filePath": "File Path",
-            "feedbackResult": "You received {{ credits }} credits for {{ part }} part",
-            "feedbackResultDetail": "with following feedback: {{ detailText }}",
+            "feedbackResult": "You received {{ credits }} credits",
```
```suggestion
            "feedbackResult": "You received {{ credits }} credit(s)",
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+            "submissionFile": "Submission Datei",
+            "notFound": "Wir haben keine neue Datei Upload Submission ohne Bewertung gefunden, bitte gehe zurück.",
+            "addFeedback": "Neues Feedback hinzufügen",
+            "assessInstruction": "Bitte klickst du auf \"Neues Feedback hinzufügen\".",
```
```suggestion
            "assessInstruction": "Bitte klicke auf \"Neues Feedback hinzufügen\".",
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+        "fileUploadAssessment": {
+            "editorTitle": "Bewertung Editor für Datei Upload Übung {{ title }}",
+            "submissionFile": "Submission Datei",
+            "notFound": "Wir haben keine neue Datei Upload Submission ohne Bewertung gefunden, bitte gehe zurück.",
```
```suggestion
            "notFound": "Wir haben keine neue Datei Upload Abgabe ohne Bewertung gefunden, bitte gehe zurück.",
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+    "artemisApp": {
+        "fileUploadAssessment": {
+            "editorTitle": "Bewertung Editor für Datei Upload Übung {{ title }}",
+            "submissionFile": "Submission Datei",
```
```suggestion
            "submissionFile": "Abgabedatei",
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+            "assessInstruction": "Bitte klickst du auf \"Neues Feedback hinzufügen\".",
+            "deleteAssessment": "Bewertung {{ index }} löschen ?",
+            "dashboard": {
+                "heading": "Tutor Dashboard zur Bewertung von Date Upload Aufgabe \"{{exerciseTitle}}\""
```
```suggestion
                "heading": "Tutor Dashboard zur Bewertung von Datei Upload Aufgabe \"{{exerciseTitle}}\""
```

lovaiible commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+            },
+            "error": {
+                "invalidAssessments": "Deine Bewertung ist nicht gültig!",
+                "invalidScoreMustBeNumber": "Das Score-Feld muss eine Zahl sein und darf nicht leer sein!",
```
```suggestion
                "invalidScoreMustBeNumber": "Das Punktzahl-Feld muss eine Zahl sein und darf nicht leer sein!",
```

lovaiible commented on src/main/webapp/i18n/en/assessment.json:
```diff
@@ -13,6 +14,38 @@
+                "submitSuccessful": "Your assessment was submitted successfully!",
+                "saveSuccessful": "Your assessment was saved successfully!",
+                "saveFailed": "Saving your assessment failed!",
+                "updateAfterComplaintSuccessful": "The Assessment was updated successfully.",
```
```suggestion
                "updateAfterComplaintSuccessful": "The assessment was updated successfully.",
```

maxr96 commented:
> 
> 
> Overall, this looks good. Nice job!
> However, there still are some bugs:
> 
>     1. After requesting more feedback and providing it as a tutor: Right after saving the additional feedback, there is a translation error in the loading animation. I couldn't take a screenshot since this happens pretty quickly.
> 
>     2. When trying to delete a participation, I receive an Http 500 response:
> 
> 
> ```
> error: Object { type: "https://www.jhipster.tech/problem/problem-with-message", title: "Internal Server Error", status: 500, … }
> ​
> headers: {…}
> ​​
> lazyInit: function lazyInit()
> ​​
> lazyUpdate: null
> ​​
> normalizedNames: Map(0)
> ​​
> <prototype>: Object { … }
> ​
> message: "Http failure response for https://artemistest.ase.in.tum.de/api/participations/150412?deleteBuildPlan=false&deleteRepository=false: 500 Internal Server Error"
> ​
> name: "HttpErrorResponse"
> ​
> ok: false
> ​
> status: 500
> ​
> statusText: "Internal Server Error"
> ​
> url: "https://artemistest.ase.in.tum.de/api/participations/150412?deleteBuildPlan=false&deleteRepository=false"
> ​
> <prototype>: {…
> ```
> 
>     1. I also couldn't delete the whole exercise afterwards (Http 500)
> 
>     2. I had one submission for my test upload exercise. After assessing it the tutor dashboard looks like this:
>        ![file_upload_info_bug](https://user-images.githubusercontent.com/27979674/65833850-1cc38d80-e2d5-11e9-89a0-4fa2ce4c4c43.png)
>        As you can see in the info box, the numbers are wrong, there is a negative value for the unassessed submissions with 100% (although all submissions are assessed), as well as a value of "0%" for the assessed submissions (although this should be 100%). The values under "Your status" also look wrong with 1/0 being 0%.

Thank you for checking the PR @ungaralex . Indeed all of the bugs that you found, I could reproduce. They should be fixed now. At least, all worked for me on the test server after the changes. 
Now, I also have an idea that for the refactoring PR we should have a method that will check if the submission/exercise is either `modeling`, `text` or `file upload`. Because if we introduce an additional exercise type similar to them, somebody will have to add a new check to all of the places.

lovaiible commented on src/main/webapp/i18n/en/assessment.json:
```diff
@@ -13,6 +14,38 @@
+                "saveSuccessful": "Your assessment was saved successfully!",
+                "saveFailed": "Saving your assessment failed!",
+                "updateAfterComplaintSuccessful": "The Assessment was updated successfully.",
+                "updateAfterComplaintFailed": "Updating the Assessment failed."
```
```suggestion
                "updateAfterComplaintFailed": "Updating the assessment failed."
```

sascha11110 commented:
I tested Alex's findings again and they now all work for me after the latest changes but I was not able to test the following:

> After requesting more feedback and providing it as a tutor: Right after saving the additional feedback, there is a translation error in the loading animation. I couldn't take a screenshot since this happens pretty quickly.

thilo-behnke commented:
> Overall, this looks good. Nice job!
> However, there still are some bugs:
> 
> 1. After requesting more feedback and providing it as a tutor: Right after saving the additional feedback, there is a translation error in the loading animation. I couldn't take a screenshot since this happens pretty quickly.
> 2. When trying to delete a participation, I receive an Http 500 response:
> 
> ```
> error: Object { type: "https://www.jhipster.tech/problem/problem-with-message", title: "Internal Server Error", status: 500, … }
> ​
> headers: {…}
> ​​
> lazyInit: function lazyInit()
> ​​
> lazyUpdate: null
> ​​
> normalizedNames: Map(0)
> ​​
> <prototype>: Object { … }
> ​
> message: "Http failure response for https://artemistest.ase.in.tum.de/api/participations/150412?deleteBuildPlan=false&deleteRepository=false: 500 Internal Server Error"
> ​
> name: "HttpErrorResponse"
> ​
> ok: false
> ​
> status: 500
> ​
> statusText: "Internal Server Error"
> ​
> url: "https://artemistest.ase.in.tum.de/api/participations/150412?deleteBuildPlan=false&deleteRepository=false"
> ​
> <prototype>: {…
> ```
> 
> 1. I also couldn't delete the whole exercise afterwards (Http 500)
> 2. I had one submission for my test upload exercise. After assessing it the tutor dashboard looks like this:
>    ![file_upload_info_bug](https://user-images.githubusercontent.com/27979674/65833850-1cc38d80-e2d5-11e9-89a0-4fa2ce4c4c43.png)
>    As you can see in the info box, the numbers are wrong, there is a negative value for the unassessed submissions with 100% (although all submissions are assessed), as well as a value of "0%" for the assessed submissions (although this should be 100%). The values under "Your status" also look wrong with 1/0 being 0%.

Not relevant for this PR, but we should write integration tests for the CRUD actions of a file upload exercise (including all expected data as e.g. participations, submissions and results). I didn't think of testing this here, so good thinking by @ungaralex.

maxr96 commented:
> 
> 
> Good job on implementing this feature! I tested it on the test server and it worked fine for me.
> I have made some suggestions on the translation strings.
> 
> Other than that I have found a few minor bugs:
> When submitting a file upload, I receive this error when I visit the `view submission` page again:
> ![image](https://user-images.githubusercontent.com/33764166/65835720-e80e0100-e2e9-11e9-9b21-f5441944b31a.png)
> Moreover I have requested `more feedback` from the assessor and there was no indication for the assessor that there was a feedback request since the number here is 0
> ![image](https://user-images.githubusercontent.com/33764166/65835740-18559f80-e2ea-11e9-81c3-368b44db6310.png)
> And one question: should the amount of feedback items be limited? Because currently I, as an assessor, can add as many feedback items as I want to.

Hey, thanks for checking it @lovaiible. I fixed the bug on `view submission` page, I forget to add the check for `this.result`. 
The issue for tutor course dashboard is not related to file upload exercise, but looks like a general issue with `TutorLeaderboardDTO`. I will check it more thoroughly on the test server and do a separate bugfix if needed.
Yes amount of feedback is not limited right now, since we don't have any reference yet.

maxr96 commented:
> 
> 
> > Overall, this looks good. Nice job!
> > However, there still are some bugs:
> > 
> > 1. After requesting more feedback and providing it as a tutor: Right after saving the additional feedback, there is a translation error in the loading animation. I couldn't take a screenshot since this happens pretty quickly.
> > 2. When trying to delete a participation, I receive an Http 500 response:
> > 
> > ```
> > error: Object { type: "https://www.jhipster.tech/problem/problem-with-message", title: "Internal Server Error", status: 500, … }
> > ​
> > headers: {…}
> > ​​
> > lazyInit: function lazyInit()
> > ​​
> > lazyUpdate: null
> > ​​
> > normalizedNames: Map(0)
> > ​​
> > <prototype>: Object { … }
> > ​
> > message: "Http failure response for https://artemistest.ase.in.tum.de/api/participations/150412?deleteBuildPlan=false&deleteRepository=false: 500 Internal Server Error"
> > ​
> > name: "HttpErrorResponse"
> > ​
> > ok: false
> > ​
> > status: 500
> > ​
> > statusText: "Internal Server Error"
> > ​
> > url: "https://artemistest.ase.in.tum.de/api/participations/150412?deleteBuildPlan=false&deleteRepository=false"
> > ​
> > <prototype>: {…
> > ```
> > 
> > 
> > 
> > 1. I also couldn't delete the whole exercise afterwards (Http 500)
> > 2. I had one submission for my test upload exercise. After assessing it the tutor dashboard looks like this:
> >    ![file_upload_info_bug](https://user-images.githubusercontent.com/27979674/65833850-1cc38d80-e2d5-11e9-89a0-4fa2ce4c4c43.png)
> >    As you can see in the info box, the numbers are wrong, there is a negative value for the unassessed submissions with 100% (although all submissions are assessed), as well as a value of "0%" for the assessed submissions (although this should be 100%). The values under "Your status" also look wrong with 1/0 being 0%.
> 
> Not relevant for this PR, but we should write integration tests for the CRUD actions of a file upload exercise (including all expected data as e.g. participations, submissions and results). I didn't think of testing this here, so good thinking by @ungaralex.

Yes, this is a good idea. I can also recheck the tests that already exists for modeling and text exercise, since they also don't cover all relevant operations.

lovaiible commented:
@maxr96 One further question: If the assessor reviews the file upload and it seems that the solution is not correct, then he cannot give 0 as score. Is that intended?

maxr96 commented:
> 
> 
> @maxr96 One further question: If the assessor reviews the file upload and it seems that the solution is not correct, then he cannot give 0 as score. Is that intended?

It is still possible, but he will need to give at least General Feedback.

maxr96 commented:
@krusche The PR is approved by two people and ready for a final review/merge.

krusche commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+                "heading": "Tutor Dashboard zur Bewertung von Datei Upload Aufgabe \"{{exerciseTitle}}\""
+            },
+            "messages": {
+                "lock": "Diese Datei kann jetzt exklusiv nur noch von dir bearbeitet werden. Bitte bewerte diese Datei bevor du ein neues öffnest.",
```
```suggestion
                "lock": "Diese Datei Abgabe kann jetzt exklusiv nur noch von dir bewertet werden. Bitte bewerte diese Datei Abgabe bevor du eine neue öffnest.",
```

krusche commented on src/main/webapp/i18n/de/fileUploadAssessment.json:
```diff
@@ -0,0 +1,24 @@
+{
+    "artemisApp": {
+        "fileUploadAssessment": {
+            "editorTitle": "Bewertung Editor für Datei Upload Übung {{ title }}",
```
```suggestion
            "editorTitle": "Bewertungseditor für Datei Upload Übung {{ title }}",
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -246,50 +253,47 @@ public FileUploadSubmissionResource(FileUploadSubmissionRepository fileUploadSub
-        if (participation.getResults().size() <= 0) {
-            List<Result> results = resultRepository.findByParticipationIdOrderByCompletionDateDesc(participation.getId());
-            participation.setResults(new HashSet<>(results));
+        // Students can only see their own models (to prevent cheating). TAs, instructors and admins can see all models.
```
```suggestion
        // Students can only see their own file uploads (to prevent cheating). TAs, instructors and admins can see all file uploads.
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -246,50 +253,47 @@ public FileUploadSubmissionResource(FileUploadSubmissionRepository fileUploadSub
+            fileUploadSubmission.setParticipation(participation);
+        }
+        else {
+            // only try to get and set the model if the fileUploadSubmission existed before
```
```suggestion
            // only try to get and set the file upload if the fileUploadSubmission existed before
```
